helix-term/src/ui/editor.rs

```diff
@@ -695,6 +750,337 @@ impl EditorView {
+                        helix_core::LineEnding::LF => 1,
+                    });
+
+                surface.set_stringn(
+                    new_x_location,
+                    additional_area.y,
+                    dots,
+                    dots.len(),
+                    theme.get("keyword.operator"),
+                );
+            }
+
+            new_offset.anchor = text.byte_to_char(node.byte_range.start);
+
+            // get all highlights from the latest point
+            let highlights = Self::doc_syntax_highlights(doc, new_offset.anchor, 1, theme);
+
+            let mut renderer = TextRenderer::new(
+                surface,
+                doc,
+                theme,
+                view.offset.horizontal_offset,
+                line_context_area,
+            );
+
+            // limit the width to its size - 1, so that it won't draw trailing whitespace characters
+            line_context_area.width = already_written - 1;
+
+            let mut text_format = doc.text_format(line_context_area.width, Some(theme));
+            text_format.soft_wrap = false;
+
+            render_text(
+                &mut renderer,
+                text,
+                new_offset,
+                &text_format,
+                &virtual_text_annotations,
+                highlights,
+                theme,
+                line_decoration,
+                translated_positions,
+            );
+
+            // next node
+            context_area.y += 1;
+        }
+    }
+
+    fn get_context_paired_range(
+        query_match: &QueryMatch,
+        start_index: u32,
+        end_index: u32,
+        top_first_byte: usize,
+        last_scan_byte: usize,
+    ) -> Option<std::ops::Range<usize>> {
+        let end_nodes: Vec<_> = query_match.nodes_for_capture_index(end_index).collect();
+        query_match
+            .nodes_for_capture_index(start_index)
+            .flat_map(move |context| {
+                end_nodes.clone().into_iter().filter_map(move |it| {
+                    let start_range = context.byte_range();
+                    let end = it.start_byte();
+                    if start_range.contains(&end)
+                            && start_range.contains(&top_first_byte)
+                            && start_range.contains(&last_scan_byte)
+                            // only match @context.params nodes that aren't at the end of the line
+                            && context.start_position().row != it.end_position().row
+                    {
+                        Some(context.start_byte()..it.end_byte().saturating_sub(1))
+                    } else {
+                        None
+                    }
+                })
+                // in some cases, the start byte of a block is on the next line
+                // which causes to show the actual first line of content instead of
+                // the actual wanted "end of signature" line
+            })
+            .next()
```

Comment:
None of the changes made address the ones fixed by my previous review: this method is still very inefficient and `.clone` is unnecessary and eager skipping can be added

Reply:
Yes I also noticed.
I havent found the time to continue working on it, sorry for that!
I thought that maybe it'd be a good idea to capture params and the return type, but I dropped the latter so I wouldn't have to change all too much on the logic, don't worry, I will incorporate your changes!

Reply:
Thanks for the contribution!
I have looked over it and adjusted it to fit the new queries, generally i prefer your code by a lot! more structure than the mess i have written.

Reply:
I'm eagerly awaiting this MR in master, thanks for working on it :)